### ISSUE TITLE
Fail to run the conformance test with 1.22

### ISSUE THREAD

**Issue by yjjy0921** [kubernetes/kubernetes#105953]
### What happened?

When I try to run conformance test with a docker command, it failed with this error output.

```bash
docker run -it k8s.gcr.io/conformance:v1.22.0
2021/10/28 02:08:23 Saving results at /tmp/results
2021/10/28 02:08:23 failed to create log file /tmp/results/e2e.log: open /tmp/results/e2e.log: no such file or directory
```

While `k8s.gcr.io/conformance:v1.20.0`  and `k8s.gcr.io/conformance:v1.21.0` are still working.
I take a look at the code `cluster/images/conformance/go-runner/main.go` (on branch release-1.22)
Seems we should create the directory for the log file before `logFile, err := os.Create(logFilePath)`

```go
func configureAndRunWithEnv(env Getenver) error {
	// Ensure we save results regardless of other errors. This helps any
	// consumer who may be polling for the results.
	resultsDir := env.Getenv(resultsDirEnvKey)
	defer saveResults(resultsDir)

	// Print the output to stdout and a logfile which will be returned
	// as part of the results tarball.
	logFilePath := filepath.Join(resultsDir, logFileName)
	logFile, err := os.Create(logFilePath)
	if err != nil {
		return errors.Wrapf(err, "failed to create log file %v", logFilePath)
	}
	mw := io.MultiWriter(os.Stdout, logFile)
	cmd := getCmd(env, mw)

	log.Printf("Running command:\n%v\n", cmdInfo(cmd))
	err = cmd.Start()
	if err != nil {
		return errors.Wrap(err, "starting command")
	}

	// Handle signals and shutdown process gracefully.
	go setupSigHandler(cmd.Process.Pid)
	return errors.Wrap(cmd.Wait(), "running command")
}
```

I fixed it with this code.
```go
        logFilePath := filepath.Join(resultsDir, logFileName)
+       if err := os.Mkdir(resultsDir, 0755); err != nil {
+               return errors.Wrapf(err, "failed to create log path %v", logFilePath)
+       }
        logFile, err := os.Create(logFilePath)
        if err != nil {
                return errors.Wrapf(err, "failed to create log file %v", logFilePath)
```

### What did you expect to happen?

The conformance test should work with v1.22.x

### How can we reproduce it (as minimally and precisely as possible)?

docker run -it k8s.gcr.io/conformance:v1.22.0

### Kubernetes version

<details>

```console
$ kubectl version
1.22.0
```

</details>

### Cloud provider

<details>
Local develop env
</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

**Comment by BenTheElder** [comment]
> docker run -it k8s.gcr.io/conformance:v1.22.0

Normally this image is invoked with some parameters specified (e.g. it will need cluster access ...), is there a use case for running it like this?

I'm curious what changed, I don't recall any related changes in 1.22

cc @johnSchnake

**Comment by yjjy0921** [comment]
> > docker run -it k8s.gcr.io/conformance:v1.22.0
> 
> Normally this image is invoked with some parameters specified (e.g. it will need cluster access ...), is there a use case for running it like this?
> 
> I'm curious what changed, I don't recall any related changes in 1.22
> 
> cc @johnSchnake

1.21 or earlier versions, Conformance test use the `cluster/images/conformance/run_e2e.sh` as the entrypoint, but now 1.22 change to the gorunner(first used as default entrypoint).

The parameter RESULTS_DIR has default value, we used to leave is as default value, and it works with 1.21 or earlier versions.

**Comment by BenTheElder** [comment]
/triage accepted
Within the project we primarily run e2e.test binary directly, this image is primarily for use-cases like sonobuoy and has been developed by that team. I'm personally way over capacity to work on this but maybe one of the sonobuoy folks can.
cc @wilsonehusin @dims https://github.com/kubernetes/kubernetes/pull/99178

**Comment by wilsonehusin** [comment]
hmm yeah I don't think running container locally was in the use case that we considered when we did #99178, because we were banking on `RESULTS_DIR` to be consistent with the mounted volume (therefore the folder must exist) like the following:

https://github.com/kubernetes/kubernetes/blob/392de8012eb4116a0467f8fbe771ba5a17e34a86/test/conformance/image/conformance-e2e.yaml#L71-L73

I think the default value is working as expected here, because the error message has the right path (i.e. `/tmp/results/e2e.log`) just that maybe the old Dockerfile has `/tmp/results` pre-initialized, which is a another option to fix this without adding Go code.

I don't feel too strongly about either approaches of fixing this, though I think the non-fixing approach will be to run the image with `RESULTS_DIR=/tmp` if you'd like to unblock yourself for now.

**Comment by johnSchnake** [comment]
Yea you normally wouldn't run the conformance image via just `docker run...`; that RESULTS_DIR it uses is referencing a mounted volume which I guess is the issue.

I don't mind at all if the code is updated to do a mkdirall call in case the folder doesnt exist.

**Comment by yjjy0921** [comment]
> Yea you normally wouldn't run the conformance image via just `docker run...`; that RESULTS_DIR it uses is referencing a mounted volume which I guess is the issue.
> 
> I don't mind at all if the code is updated to do a mkdirall call in case the folder doesnt exist.

Okay, let me create a PR to do the mkdir if the foler is not exists

**Comment by dims** [comment]
@yjjy0921 did you create a PR yet?

**Comment by yjjy0921** [comment]
> PR

yes, please help to review the following PR

**Comment by johnSchnake** [comment]
This actually just came up and _helped_ me find a bug elsewhere. I was able to just mount the pwd as the results directory (which is an alternative to fixing this at all). In my case, I actually didn't expect to need that at all, so I realized the entrypoint for the conformance image had changed and I had to invoke the image with the `--entrypoint=...` option before specifying the image name.

**Comment by dims** [comment]
closing as WONTFIX
